### ISSUE TITLE
"Unable to sync" steam error while using multi-instance commands

### ISSUE THREAD

**Issue by kopiezzo** [DuoStream/Duo#89]
Hey, I'm experiencing issues with Steam multi-instance. Here's my setup:

Host: Launching Steam with a batch file:
`start "Steam_HOST" "C:\Program Files (x86)\Steam\steam.exe" -master_ipc_name_override HOST -userchooser HOST`

Client: Launching Steam with default command:
`cmd /c "steam.exe" -master_ipc_name_override LK -userchooser LK -gamepadui`

Now, whenever I have Steam open on the HOST, then I log in to the client to play on a phone, etc., and everything works. But when I stop using my phone and want to play on the HOST again, I'm getting an "Unable to sync" error, and I have to restart Steam on the HOST every time. Is that expected?

![error](https://github.com/DuoStream/Duo/assets/15052907/3ad51fa8-d6a9-496c-9b66-6607c65ee626)

**Comment by Black-Seraph** [comment]
Did you set the environment variables needed for multi mode?

**Comment by samkitty** [comment]
This issue isn't related to Duo as far as I can tell. It's caused by Steam Cloud saves and it can effect anyone who uses the same Steam account to play on multiple devices. You can see many Steam Deck users complain about sync issues when Steam Cloud is enabled. I currently have my client play in offline mode to avoid the headache of Steam Cloud, but I'm actively looking for a robust fix. Duo is unique because our game saves are on the same PC but under a different user, so maybe using a symlink we can automatically transfer the most recent game save data between the 2 users to avoid Steam Cloud sync issues.

**Comment by Black-Seraph** [comment]
@samkitty I fully agree with you there but will leave the issue open for a little while longer anyway to foster discussion of potential workarounds.

If a workaround is found that is somewhat automatable then I'd be more than willing to implement it into Duo.

**Comment by Black-Seraph** [comment]
> Thank you for the explanation, so for now the only workaround to keep playing in online mode is to close steam on host every time, correct?

That, or run multiple Steam accounts, which, for me, is a valid option as Duo is being used to share a single computer with my wife.

**Comment by samkitty** [comment]
> So for now the only workaround to keep playing in online mode is to close steam on host every time, correct?
Disabling Cloud Saves in Steam is also an option, but that will prevent some of your game saves from syncing between host<>client.

There are a few tools you can use to sync your game saves without relying on Steam but I honestly haven't tried any of them yet.
- [Open Cloud Saves](https://github.com/DavidDeSimone/OpenCloudSaves)
- [GameSave Manager](https://www.gamesave-manager.com/)
- [Ludusavi](https://github.com/mtkennerly/ludusavi)
- [Ludusavi plugin for Playnite](https://github.com/mtkennerly/ludusavi-playnite)

Another option, unique for Duo users is to create a symlink between the game save folders on each of the two windows users. That way, any change that happens on either user's files should sync automatically with the second user and Steam will always have the most recent files so no sync issues should occur. Keep in mind I have not tested this theory and can't say for sure if it will work correctly. This tool makes creating symlinks easy: 
- [Symlink Creator](https://github.com/arnobpl/SymlinkCreator)

And you can find the location of the relevant save files on your PC using [PCGamingWiki](https://www.pcgamingwiki.com/wiki/Home), just search for the game and scroll down to **Save Game Data Location**.

**Comment by Black-Seraph** [comment]
I personally have moved all of my stuff into Syncthing.

That one works surprisingly well for my needs.

**Comment by kopiezzo** [comment]
> I personally have moved all of my stuff into Syncthing.
That one works surprisingly well for my needs.

I am already using Syncthing on my server and phone. Could you recommend which folder to sync between all devices to synchronize all saved games at once and disable cloud sync altogether?

**Comment by Black-Seraph** [comment]
- **C:\Users\UserName\AppData\Local** (with a few ignore patterns for Windows folders)
- **C:\Users\UserName\Documents** (with a few ignore patterns based on what you usually store in your Documents folder)
- **C:\Users\UserName\Saved Games**
- **C:\Users\Public\Documents** (with a few ignore patterns if you're making use of PowerShell modules or Hyper-V)

All folders should be setup with a **desktop.ini** ignore pattern as Windows loves to plaster these in pretty much all folders.

**Comment by kopiezzo** [comment]
With your advice, I managed to handle the previous errors, set up Syncthing for the folders you indicated, and it seems to be working. However, I got stuck with another error. My current setup is as follows:

**HOST: "C:\Program Files (x86)\Steam\steam.exe" -master_ipc_name_override HOST**
(I also tried "C:\Program Files (x86)\Steam\steam.exe" -master_ipc_name_override HOST -userchooser HOST)

**CLIENT: cmd /c "steam.exe" -master_ipc_name_override LK -userchooser LK -gamepadui**

On the HOST account, Steam is running all the time, and the process looks as follows:

1. Steam Cloud is disabled.
2. Launch the game on the host = works
3. Launch the game on the client after launching the client = works
4. Close the client
5. Try to launch the game again on the host = game is launching but shows a new error when trying to log in (screenshot attached)
6. Restart Steam on the host = the game works again

The tested game is Last Epoch in online mode. Is there a way to resolve this error as well? 
I apologize for dragging this on, but I hope it might help other users in the future too.

![lastepoch](https://github.com/DuoStream/Duo/assets/15052907/22ff6a59-2c9b-498a-a571-72574ea249b3)

**Comment by samkitty** [comment]
Look at the bottom of your steam client library, does it say **No Connection**?
![AJxyN](https://github.com/DuoStream/Duo/assets/57629730/3bc62c88-5e22-4887-b884-cdc302c3736e)
If so then Steam is forcing the client into offline mode because it is detecting another instance of the same account is online **on the same IP address**, even if no game is currently running on that other instance.

Try this `cmd /c "steam.exe" -private_ip_override 192.0.0.0 -master_ipc_name_override LK -userchooser LK -gamepadui`

**Comment by samkitty** [comment]
@Black-Seraph A somewhat silly solution I found for the issue in OP is to use `runas /user:` command to launch steam.exe under the host user from within the client. That way if you open task manager you'll see games are being launched under Host user while showing on the client display. Any game running that way will simply use the same save data and any relevant game files as the host. This fixes all sync issues and doesn't seem to have any downsides (aside from having to type host user password every time you want to launch Steam).
![runas-cmd](https://github.com/DuoStream/Duo/assets/57629730/19a87df0-9602-4af9-a62a-6faee46e89d2)

**Comment by Black-Seraph** [comment]
@samkitty Your runas solution is great, but I think this is something that should be left up to the user to set up and use as there's no reliable way to set this up programmatically within Duo.

Your private_ip_override is a very nice find though, one I am tempted to add to Duo's default Steam shortcut as this one **can** be automated by iterating over the local network adapter IP addresses using the Win32 API.

That would solve @kopiezzo's problem with Last Epoch.

**Comment by Black-Seraph** [comment]
All of these workarounds should no longer be required starting at version 1.5.1.
